File: lab1/1_2.cpp
#include <iostream>
#include <cmath>
#include <functional>

const double EPSILON = 1e-6;

// Функция
double f(double x) {
    return std::log(x) + std::pow((x + 1), 3);
}

// Производная функции для метода Ньютона
double f_derivative(double x) {
    return 1/x + 3*std::pow((x + 1), 2);
}

// Метод дихотомии (бисекции)
double bisectionMethod(double a, double b) {
    if (f(a) * f(b) >= 0) {
        std::cerr << "Invalid interval for the bisection method." << std::endl;
        return -1;
    }

    double c;
    while ((b - a) >= EPSILON) {
        c = (a + b) / 2;
        if (f(c) == 0.0) {
            break;
        } else if (f(c) * f(a) < 0) {
            b = c;
        } else {
            a = c;
        }
    }
    return c;
}

// Функция для метода простых итераций
double g(double x) {
    return -std::log(x) / std::pow((x + 1), 3);
}

// Метод простых итераций
double fixedPointIteration(double x0) {
    double x1;
    do {
        x1 = g(x0);
        if (std::abs(x1 - x0) < EPSILON) {
            break;
        }
        x0 = x1;
    } while (true);
    return x1;
}

// Метод хорд (секущих)
double secantMethod(double a, double b) {
    double c;
    while (std::abs(b - a) >= EPSILON) {
        c = b - (f(b) * (b - a)) / (f(b) - f(a));
        a = b;
        b = c;
    }
    return c;
}

// Метод Ньютона
double newtonMethod(double x0) {
    double x1;
    while (true) {
        x1 = x0 - f(x0) / f_derivative(x0);
        if (std::abs(x1 - x0) < EPSILON) {
            break;
        }
        x0 = x1;
    }
    return x1;
}

int main() {
    double a = 0.1, b = 1.0; // Интервал для метода дихотомии и хорд
    double initial_guess = 0.5; // Начальное приближение для метода простых итераций и Ньютона

    // Метод дихотомии
    double root_bisection = bisectionMethod(a, b);
    std::cout << "Корень, найденный методом дихотомии: " << root_bisection << std::endl;

    // Метод простых итераций
    double root_fixed_point = fixedPointIteration(initial_guess);
    std::cout << "Корень, найденный методом простых итераций: " << root_fixed_point << std::endl;

    // Метод хорд
    double root_secant = secantMethod(a, b);
    std::cout << "Корень, найденный методом хорд: " << root_secant << std::endl;

    // Метод Ньютона
    double root_newton = newtonMethod(initial_guess);
    std::cout << "Корень, найденный методом Ньютона: " << root_newton << std::endl;

    return 0;
}

File: lab1/1_2.py
import math

# Заданное уравнение
def f(x):
    return math.log(x) + (x + 1) ** 3

# Производная функции для метода Ньютона
def f_derivative(x):
    return 1 / x + 3 * (x + 1) ** 2

# Метод дихотомии (бисекции)
def bisection_method(a, b, epsilon=1e-6):
    if f(a) * f(b) >= 0:
        raise ValueError("Invalid interval for the bisection method.")
    
    while (b - a) / 2.0 > epsilon:
        midpoint = (a + b) / 2.0
        if f(midpoint) == 0:
            return midpoint
        elif f(a) * f(midpoint) < 0:
            b = midpoint
        else:
            a = midpoint
    
    return (a + b) / 2.0

# Функция для метода простых итераций
def g(x):
    return -math.log(x) / ((x + 1) ** 3)

# Метод простых итераций
def fixed_point_iteration(x0, epsilon=1e-6):
    while True:
        x1 = g(x0)
        if abs(x1 - x0) < epsilon:
            break
        x0 = x1
    return x1

# Метод хорд (секущих)
def secant_method(a, b, epsilon=1e-6):
    while abs(b - a) >= epsilon:
        c = b - (f(b) * (b - a)) / (f(b) - f(a))
        a, b = b, c
    return c

# Метод Ньютона
def newton_method(x0, epsilon=1e-6):
    while True:
        x1 = x0 - f(x0) / f_derivative(x0)
        if abs(x1 - x0) < epsilon:
            break
        x0 = x1
    return x1

if __name__ == "__main__":
    # Интервал для метода дихотомии и хорд
    a, b = 0.1, 1.0
    # Начальное приближение для метода простых итераций и Ньютона
    initial_guess = 0.5

    # Метод дихотомии
    root_bisection = bisection_method(a, b)
    print(f"Корень, найденный методом дихотомии: {root_bisection}")

    # Метод простых итераций
    root_fixed_point = fixed_point_iteration(initial_guess)
    print(f"Корень, найденный методом простых итераций: {root_fixed_point}")

    # Метод хорд
    root_secant = secant_method(a, b)
    print(f"Корень, найденный методом хорд: {root_secant}")

    # Метод Ньютона
    root_newton = newton_method(initial_guess)
    print(f"Корень, найденный методом Ньютона: {root_newton}")
